akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -1929,6 +1929,31 @@ trait FlowOps[+Out, +Mat] {
   def prefixAndTail[U >: Out](n: Int): Repr[(immutable.Seq[Out], Source[U, NotUsed])] =
     via(new PrefixAndTail[Out](n))
 
+  /**
+   * Takes up to `n` elements from the stream (less than `n` only if the upstream completes before emitting `n` elements),
+   * then apply `f` on these elements in order to obtain a flow, this flow is then materialized and the rest of the input is processed by this flow (similar to via ).
+   * This method returns a flow consuming the rest of the stream producing the materialized flow's output.
+   *
+   * '''Emits when''' the materialized flow emits.
+   *  Notice the first `n` elements are buffered internally before materializing the flow, This flow will then be materialized and connected to the rest of the upstream - producing elements at its own discretion (might 'swallow' or multiply elements).
+   *
+   * '''Backpressures when''' the materialized flow backpressures
+   *
+   * '''Completes when''' the materialized flow completes.
+   *  If upstream completes before producing `n` elements, `f` will be applied with the provided elements,
+   *  the resulting flow will be materialized and signalled for upstream completion, it can then cancel/complete at its own discretion.
+   *
+   * '''Cancels when''' upstream cancels, application of `f` fails, materialization of the resulting flow fails or the materialized flow cancels.
+   *  Notice the materialized flow may apply its own logic in case of upstream or downstream cancellations.
+   *
+   *  @param n the number of elements to accumulate before materializing the downstream flow.
+   *  @param f a function that produces the downstream flow based on the upstream's prefix.
+   **/
+  def prefixAndDownstream[Out2, Mat2](n : Int)(f : immutable.Seq[Out] => Flow[Out, Out2, Mat2]): Repr[Out2] = {
```

Comment:
I like the signature, and having a separate Mat version is good.

Some naming consideration though, I still think it should be a `flatten` since that is much what it does, `flatMapPrefixAndTail` is a mouthful but I think it best conveys what the operator does.

Reply:
I think this operator is somewhat of a dynamic version of via/viaMat.
notice there's no flattening here, we just connect two flows together when one of them is not known up-front.

having that said, I think you as part of the akka team should decide about this operator's name.

Reply:
Yeah, it's not quite `flatMap` either, true, hmmm. I'll talk to the team and see if anyone else has a good idea.

Reply:
@ennru and me had a discussion and came up the suggestion `fromPrefix` which good since it says exactly what it does: creates the flow from a prefix, not so nice inside a stream flow api definition though `source.map(yadi).fromPrefix(2)(prefix => flow)`. Perhaps `flowFromPrefix` to cover that case more nicely.

Enno also suggested something like `peek` since it peeks into the stream, but as such operations (queue for example) usually do not consume elements which this operator does.

Reply:
I should add that making it something like `via` is not good name because that is not an operator so very different from this. This is somewhat alike `lazyFlow(() => Flow)` in that it delays creation (but further), that doesn't help much with naming ideas for me though.

Reply:
Discussed this a bit back and forth with @jrudolph and I think we both ended up liking `flatMapPrefix` best. Seeing that both standalone and in a flow I think gives a good intuition for what it does if you have used any other monady APIs (like the Scala collections):

```scala
// flow api
source.map(yadi).flatMapPrefix(2)(prefix => flow).to(sink)

// standalone
val flow = Flow.flatMapPrefix(2)(prefix => flow)
```

It's not perfect, since in the flow api the left hand side value is a `Source`, and the flatten implies `A => C[A]`, however I think that is ok since such a signature wouldn't anyway make any sense and we can think of the `C` as stream instead of specifically `Source`.